python/tvm/micro/base.py

```diff
@@ -0,0 +1,178 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Base definitions for micro."""
+
+from __future__ import absolute_import
+
+import logging
+import os
+
+import tvm.module
+from tvm.contrib import graph_runtime, util
+from tvm import relay
+
+from .._ffi.function import _init_api
+from .._ffi.libinfo import find_include_path
+from .cross_compile import create_lib
+
+SUPPORTED_DEVICE_TYPES = ["host", "openocd"]
+
+class Session:
+    """MicroTVM Session
+
+    Example
+    --------
+    .. code-block:: python
+
+      c_mod = ...  # some module generated with "c" as the target
+      device_type = "host"
+      with tvm.micro.Session(device_type) as sess:
+          sess.create_micro_mod(c_mod)
+    """
+
+    def __init__(self, device_type, binutil_prefix, port=0):
+        """Stores parameters for initializing a micro device session.
+
+        The session is not initialized until the constructed object is used
+        in a `with` block.
+
+        Parameters
+        ----------
+        device_type : str
+            type of low-level device
+
+        binutil_prefix : str
+            binutil prefix to be used. For example, a prefix of
+            "riscv64-unknown-elf-" means "riscv64-unknown-elf-gcc" is used as
+            the compiler and "riscv64-unknown-elf-ld" is used as the linker,
+            etc.
+
+        port : integer, optional
+            port number of OpenOCD server
+        """
+        if device_type not in SUPPORTED_DEVICE_TYPES:
+            raise RuntimeError("unknown micro device type \"{}\"".format(device_type))
+
+        self.device_type = device_type
+        self.binutil_prefix = binutil_prefix
+        self.port = port
+
+    def micro_build(self, func: relay.Function, params={}):
```

Comment:
keep this function as global as it is not dependent on session.  We should minimize the number of functions, In this case. this function should simply be removed

Reply:
Also change `params` default value from `{}` to `None`.

https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python/26320938